In [57]:
import sqlalchemy

In [58]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [59]:
%sql postgresql://postgres:***@127.0.0.1:5432/music

In [60]:
%%sql

SELECT table_name, table_schema
    FROM information_schema.tables
    WHERE table_schema NOT IN ('pg_catalog', 'information_schema');

 * postgresql://postgres:***@127.0.0.1:5432/music
11 rows affected.


[('Artist', 'public'),
 ('Album', 'public'),
 ('Employee', 'public'),
 ('Customer', 'public'),
 ('Invoice', 'public'),
 ('InvoiceLine', 'public'),
 ('Track', 'public'),
 ('Playlist', 'public'),
 ('PlaylistTrack', 'public'),
 ('Genre', 'public'),
 ('MediaType', 'public')]

# Who's the composer of the track 'Cochise'?

In [61]:
%%sql

SELECT *
    FROM information_schema.columns
    WHERE table_name = 'Track';

 * postgresql://postgres:***@127.0.0.1:5432/music
9 rows affected.


[('music', 'public', 'Track', 'UnitPrice', 9, None, 'NO', 'numeric', None, None, 10, 10, 2, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'numeric', None, None, None, None, '9', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Track', 'Bytes', 8, None, 'YES', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'int4', None, None, None, None, '8', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Track', 'TrackId', 1, None, 'NO', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'int4', None, None, None, None, '1', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Track', 'AlbumId', 3, None, 'YES', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'int4', None, None, None, None, '3', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Track', 'MediaTypeId', 4, None, 'NO', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'int4', None, None, None, None, '4', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Track', 'GenreId', 5, None, 'YES', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'int4', None, None, None, None, '5', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Track', 'Milliseconds', 7, None, 'NO', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'int4', None, None, None, None, '7', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Track', 'Name', 2, None, 'NO', 'character varying', 200, 800, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'varchar', None, None, None, None, '2', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Track', 'Composer', 6, None, 'YES', 'character varying', 220, 880, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'varchar', None, None, None, None, '6', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES')]

In [62]:
%%sql

SELECT "Name", "Composer"
    FROM "Track"
    WHERE "Name" = 'Cochise';

 * postgresql://postgres:***@127.0.0.1:5432/music
1 rows affected.


[('Cochise', 'Audioslave/Chris Cornell')]

Answer: Audioslave/Chris Cornell

# What's the name of genre for the track 'Cochise'?

In [63]:
%%sql

SELECT *
    FROM information_schema.columns
    WHERE table_name = 'Genre';

 * postgresql://postgres:***@127.0.0.1:5432/music
2 rows affected.


[('music', 'public', 'Genre', 'GenreId', 1, None, 'NO', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'int4', None, None, None, None, '1', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Genre', 'Name', 2, None, 'YES', 'character varying', 120, 480, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'varchar', None, None, None, None, '2', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES')]

In [64]:
%%sql

SELECT T."Name" AS "TrackName", G."Name" AS "GenreName"
    FROM "Track" T, "Genre" G
    WHERE T."GenreId" = G."GenreId" AND T."Name" = 'Cochise';

 * postgresql://postgres:***@127.0.0.1:5432/music
1 rows affected.


[('Cochise', 'Rock')]

Answer: Rock

# What's the top 3 names of genre with the most tracks? How many tracks are there total in that genre?

In [65]:
%%sql

SELECT G."Name" AS "GenreName", COUNT(G."Name") AS "NumOfTracks" 
    FROM "Track" T
    JOIN "Genre" G
    ON T."GenreId" = G."GenreId"
    GROUP BY G."Name"
    ORDER BY Count(G."Name") DESC
    LIMIT 3;

 * postgresql://postgres:***@127.0.0.1:5432/music
3 rows affected.


[('Rock', 1259), ('Metal', 342), ('Alternative & Punk', 265)]

Answer: The top 3 names of genre with the most tracks are Rock, Metal, Alternative & Punk, and their tracks are 1259, 342, 265 separately.

# What's the name of the genre with the 3rd most tracks? How many tracks are there total in that genre?

In [66]:
%%sql

SELECT G."Name" AS "GenreName", COUNT(G."Name") AS "NumOfTracks" 
    FROM "Track" T
    JOIN "Genre" G
    ON T."GenreId" = G."GenreId"
    GROUP BY G."Name"
    ORDER BY Count(G."Name") DESC
    LIMIT 1 OFFSET 2;

 * postgresql://postgres:***@127.0.0.1:5432/music
1 rows affected.


[('Alternative & Punk', 265)]

Answer: The name of genre with the 3rd most tracks is Alternative & Punk, and its number of tracks is 265.

# How many tracks does the artist 'Metallica' have?

In [67]:
%%sql

SELECT R."Name" AS "ArtistName", COUNT(R."Name") AS "NumOfTracks"
    FROM "Track" T
    JOIN "Album" A
    ON T."AlbumId" = A."AlbumId"
    JOIN "Artist" R
    ON A."ArtistId" = R."ArtistId"
    GROUP BY R."Name"
    HAVING R."Name" = 'Metallica';

 * postgresql://postgres:***@127.0.0.1:5432/music
1 rows affected.


[('Metallica', 112)]

Answer: 112

# Which Country buys the most music (invoice)?

In [68]:
%%sql

SELECT *
    FROM information_schema.columns
    WHERE table_name = 'Invoice';

 * postgresql://postgres:***@127.0.0.1:5432/music
9 rows affected.


[('music', 'public', 'Invoice', 'InvoiceId', 1, None, 'NO', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'int4', None, None, None, None, '1', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Invoice', 'CustomerId', 2, None, 'NO', 'integer', None, None, 32, 2, 0, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'int4', None, None, None, None, '2', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Invoice', 'InvoiceDate', 3, None, 'NO', 'timestamp without time zone', None, None, None, None, None, 6, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'timestamp', None, None, None, None, '3', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Invoice', 'Total', 9, None, 'NO', 'numeric', None, None, 10, 10, 2, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'numeric', None, None, None, None, '9', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Invoice', 'BillingCity', 5, None, 'YES', 'character varying', 40, 160, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'varchar', None, None, None, None, '5', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Invoice', 'BillingState', 6, None, 'YES', 'character varying', 40, 160, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'varchar', None, None, None, None, '6', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Invoice', 'BillingCountry', 7, None, 'YES', 'character varying', 40, 160, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'varchar', None, None, None, None, '7', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Invoice', 'BillingPostalCode', 8, None, 'YES', 'character varying', 10, 40, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'varchar', None, None, None, None, '8', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES'),
 ('music', 'public', 'Invoice', 'BillingAddress', 4, None, 'YES', 'character varying', 70, 280, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'music', 'pg_catalog', 'varchar', None, None, None, None, '4', 'NO', 'NO', None, None, None, None, None, 'NO', 'NEVER', None, 'YES')]

In [69]:
%%sql

SELECT "BillingCountry", COUNT("BillingCountry") AS "NumOfInvoices"
    FROM "Invoice"
    GROUP BY "BillingCountry"
    ORDER BY COUNT("BillingCountry") DESC
    LIMIT 1;

 * postgresql://postgres:***@127.0.0.1:5432/music
1 rows affected.


[('USA', 91)]

Answer: USA

# What are the top 5 Genres in the US?

In [70]:
%%sql

SELECT G."Name" AS "GenreName", COUNT(G."Name") AS "NumOfInvoice"
    FROM "InvoiceLine" IL, "Invoice" I, "Track" T, "Genre" G
    WHERE IL."InvoiceId" = I."InvoiceId" 
    AND IL."TrackId" = T."TrackId" 
    AND T."GenreId" = G."GenreId" 
    AND I."BillingCountry" = 'USA'
    GROUP BY G."Name"
    ORDER BY COUNT(G."Name") DESC
    LIMIT 5;

 * postgresql://postgres:***@127.0.0.1:5432/music
5 rows affected.


[('Rock', 157),
 ('Metal', 63),
 ('Latin', 41),
 ('Alternative & Punk', 41),
 ('Jazz', 19)]

Answer: The top 5 genres in the US is Rock, Metal, Latin, Alternative & Punk, Jazz.